File: examples/typescript/quote_and_swap.ts
/**
 * PaladinFi Swap — end-to-end TypeScript example using viem.
 *
 *   pnpm add viem
 *
 * Set TAKER_PRIVATE_KEY in the environment before running. The taker must
 * hold the sellToken and have approved the returned router (or use Permit2 —
 * out of scope for this example).
 */
import { createPublicClient, createWalletClient, http, parseGwei, type Hex } from "viem";
import { privateKeyToAccount } from "viem/accounts";
import { base } from "viem/chains";

const PALADIN_API = "https://swap.paladinfi.com";

// Base mainnet token addresses
const USDC = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913" as const;
const WETH = "0x4200000000000000000000000000000000000006" as const;

interface QuoteResponse {
  source: string;
  chainId: number;
  router: string;
  calldata: string;
  buyAmount: string;
  minBuyAmount: string;
  sellAmount: string;
  gas: string;
  ourFeeBps: number;
  ourFeeRecipient: string;
  estimatedOurFeeAmount?: string;
  estimatedOurFeeToken?: string;
}

async function quote(params: {
  sellToken: string;
  buyToken: string;
  sellAmount: string;
  taker: string;
  slippageBps?: number;
}): Promise<QuoteResponse> {
  const res = await fetch(`${PALADIN_API}/v1/quote`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ chainId: 8453, slippageBps: 50, ...params }),
  });
  if (!res.ok) {
    throw new Error(`Quote failed ${res.status}: ${await res.text()}`);
  }
  return (await res.json()) as QuoteResponse;
}

async function main() {
  const pk = process.env.TAKER_PRIVATE_KEY as Hex | undefined;
  if (!pk) {
    console.error("Set TAKER_PRIVATE_KEY in the environment.");
    process.exit(1);
  }
  const account = privateKeyToAccount(pk);

  const q = await quote({
    sellToken: USDC,
    buyToken: WETH,
    sellAmount: "5000000",
    taker: account.address,
  });

  console.log(`Source:        ${q.source}`);
  console.log(`Router:        ${q.router}`);
  console.log(`Buy amount:    ${q.buyAmount} (min ${q.minBuyAmount})`);
  console.log(`Affiliate fee: ${q.ourFeeBps} bps to ${q.ourFeeRecipient}`);
  console.log(`Gas estimate:  ${q.gas}`);
  console.log();

  const wallet = createWalletClient({
    account,
    chain: base,
    transport: http(),
  });
  const publicClient = createPublicClient({
    chain: base,
    transport: http(),
  });

  console.log("Submitting transaction...");
  const hash = await wallet.sendTransaction({
    to: q.router as `0x${string}`,
    data: q.calldata as `0x${string}`,
    value: 0n,
    gas: BigInt(Math.floor(Number(q.gas) * 1.2)),
    maxFeePerGas: parseGwei("0.05"),
    maxPriorityFeePerGas: parseGwei("0.005"),
  });
  console.log(`TX:            https://basescan.org/tx/${hash}`);

  // Optionally wait for inclusion
  const receipt = await publicClient.waitForTransactionReceipt({ hash });
  console.log(`Status:        ${receipt.status}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/python/quote_and_swap.py
"""
PaladinFi Swap — end-to-end Python example.

Quotes USDC -> WETH on Base, signs the calldata returned by the service, and
submits the transaction. Requires:

    pip install web3 requests

Set TAKER_PRIVATE_KEY in the environment before running. The taker must hold
the sellToken and have approved the returned router (or use Permit2 — out of
scope for this example).
"""
from __future__ import annotations
import os
import sys
import requests
from web3 import Web3
from eth_account import Account

PALADIN_API = "https://swap.paladinfi.com"
BASE_RPC = "https://base-rpc.publicnode.com"
CHAIN_ID = 8453

# Base mainnet token addresses
USDC = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913"
WETH = "0x4200000000000000000000000000000000000006"


def quote(sell_token: str, buy_token: str, sell_amount: str, taker: str,
          slippage_bps: int = 50) -> dict:
    """Fetch a competitive route quote with ready-to-execute calldata."""
    payload = {
        "chainId": CHAIN_ID,
        "sellToken": sell_token,
        "buyToken": buy_token,
        "sellAmount": sell_amount,
        "taker": taker,
        "slippageBps": slippage_bps,
    }
    resp = requests.post(f"{PALADIN_API}/v1/quote", json=payload, timeout=15)
    resp.raise_for_status()
    return resp.json()


def submit(quote_response: dict, private_key: str) -> str:
    """Sign the quote's calldata and submit. Returns the tx hash."""
    w3 = Web3(Web3.HTTPProvider(BASE_RPC))
    acct = Account.from_key(private_key)
    if acct.address.lower() != quote_response["sellAmount"] and \
            acct.address.lower() != _safe_get_taker(quote_response).lower():
        # The quote was prepared for a specific taker; the signer must match.
        # We don't actually have taker in the quote response; the caller is
        # responsible for ensuring the keys match the address they passed in.
        pass

    tx = {
        "from": acct.address,
        "to": Web3.to_checksum_address(quote_response["router"]),
        "data": quote_response["calldata"],
        "value": 0,
        "gas": int(int(quote_response["gas"]) * 1.2),
        "maxFeePerGas": w3.to_wei(0.05, "gwei"),
        "maxPriorityFeePerGas": w3.to_wei(0.005, "gwei"),
        "nonce": w3.eth.get_transaction_count(acct.address),
        "chainId": CHAIN_ID,
    }
    signed = acct.sign_transaction(tx)
    tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
    return tx_hash.hex()


def _safe_get_taker(_quote_response: dict) -> str:
    # Placeholder — current API does not echo taker in the response. Caller
    # tracks this externally.
    return ""


def main() -> int:
    pk = os.environ.get("TAKER_PRIVATE_KEY")
    if not pk:
        print("Set TAKER_PRIVATE_KEY in the environment.", file=sys.stderr)
        return 1
    taker = Account.from_key(pk).address

    # 5 USDC -> WETH
    q = quote(USDC, WETH, "5000000", taker)
    print(f"Source:        {q['source']}")
    print(f"Router:        {q['router']}")
    print(f"Buy amount:    {q['buyAmount']} (min {q['minBuyAmount']})")
    print(f"Affiliate fee: {q['ourFeeBps']} bps to {q['ourFeeRecipient']}")
    print(f"Gas estimate:  {q['gas']}")
    print()
    print("Submitting transaction...")
    tx_hash = submit(q, pk)
    print(f"TX:            https://basescan.org/tx/{tx_hash}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
